includes/utils/class-amp-dom-document.php

```diff
@@ -0,0 +1,313 @@
+	}
+
+	/**
+	 * Load HTML from a string.
+	 *
+	 * @link  https://php.net/manual/domdocument.loadhtml.php
+	 *
+	 * @param string     $source  The HTML string.
+	 * @param int|string $options Optional. Specify additional Libxml parameters.
+	 * @return bool true on success or false on failure.
+	 */
+	public function loadHTML( $source, $options = 0 ) {
+		$source = $this->maybe_add_head_or_body( $source );
+
+		$this->original_encoding = $this->detect_and_strip_encoding( $source );
+
+		if ( self::AMP_ENCODING !== strtolower( $this->original_encoding ) ) {
+			$source = $this->adapt_encoding( $source );
+		}
+
+		// Force-add http-equiv charset to make DOMDocument behave as it should.
+		// See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+		$source = str_replace(
+			'<head>',
+			'<head><meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">',
+			$source
+		);
+
+		$success = parent::loadHTML( $source, $options );
+
+		if ( $success ) {
+			$this->encoding = self::AMP_ENCODING;
+			$head           = $this->getElementsByTagName( 'head' )->item( 0 );
+			$head->removeChild( $head->firstChild );
+		}
+
+		return $success;
+	}
+
+	/**
+	 * Dumps the internal document into a string using HTML formatting.
+	 *
+	 * @link  https://php.net/manual/domdocument.savehtml.php
+	 *
+	 * @param DOMNode $node Optional. Parameter to output a subset of the document.
+	 * @return string The HTML, or false if an error occurred.
+	 */
+	public function saveHTML( DOMNode $node = null ) {
+		// Force-add http-equiv charset to make DOMDocument behave as it should.
+		// See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+		$head    = $this->getElementsByTagName( 'head' )->item( 0 );
+		$charset = AMP_DOM_Utils::create_node(
+			$this,
+			'meta',
+			[
+				'http-equiv' => 'content-type',
+				'content'    => 'text/html; charset=' . self::AMP_ENCODING,
+			]
+		);
+		$head->insertBefore( $charset, $head->firstChild );
+
+		return str_replace( '<meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">', '', parent::saveHTML( $node ) );
+	}
+
+	/**
+	 * Maybe add the <head> and/or <body> tag(s).
+	 *
+	 * @param string $content Content to add the head or body tags to.
+	 * @return string Adapted content.
+	 */
+	private function maybe_add_head_or_body( $content ) {
+		$substring = substr( $content, 0, 5000 );
+		if ( false === strpos( $substring, '<body' ) ) {
+			if ( false === strpos( $substring, '<head>' ) ) {
+				// Create the required HTML structure if none exists yet.
+				$content = "<html><head></head><body>{$content}</body></html>";
+			} else {
+				// <head> seems to be present without <body>.
+				$content = preg_replace( '#</head>(.*)</html>#', '</head><body>$1</body>', $content );
+			}
+		} elseif ( false === strpos( $substring, '<head>' ) ) {
+			// Create a <head> element if none exists yet.
+			$content = str_replace( '<body', '<head></head><body', $content );
```

Comment:
What if `<head>` is present but it contains attributes, like:

```html
<head profile='http://www.acme.com/profiles/core'>
```

Reply:
Yes, will change to a `preg_replace()`.